### PR TITLE
Faster TraceBuffer for CRuby

### PR DESCRIPTION
**Pull request by marcotc**
**_tl;dr: reduction of 17-93% of allocated bytes, 5-15% faster._**

Leveraging the fact that the native Ruby `Array` ["is thread-safe in practice because CRuby runs threads one at a time and does not do context switching during the execution of C functions"](https://github.com/ruby-concurrency/concurrent-ruby/blob/c1114a0c6891d9634f019f1f9fe58dcae8658964/lib/concurrent-ruby/concurrent/array.rb#L23-L27), this PR implements a version of the `TraceBuffer` that does not use explicit locking.

The buffer is one of the tracer hot-stops, being a sync point between the application critical path and the tracer's worker thread. All traces will eventually be pushed into the buffer, so improvements to it affect all instrumentations.

This version works correctly on CRuby, but will not maintain the same guarantees under other runtimes, like JRuby. For this reason, we kept the existing implementation, which utilizes explicit locking, for non-CRuby environments.

The benchmarks below (also included in the PR) use the default buffer size of 1000 traces.
When pushing over 1000 traces into the buffer, our fair eviction policy will take place. The 2000 traces benchmark covers this case. Increasing the number of traces pushed even more yielded the same performance results, so we stop at 2000.

The reduction in memory usage is the most notable improvement, with memory usage being constant for the new implementation:
```
Before(ThreadSafeBuffer) [bytes allocated (objects created)]
    10 traces:   91840.00 (   1900.00)
   100 traces:  177376.00 (   1900.00)
  1000 traces: 1244200.00 (   1900.00)
  2000 traces: 1244200.00 (   1900.00)

After(CRubyTraceBuffer)  [bytes allocated (objects created)]
    10 traces:   76000.00 (   1900.00)
   100 traces:   76000.00 (   1900.00)
  1000 traces:   76000.00 (   1900.00)
  2000 traces:   76000.00 (   1900.00)

Comparison (% reduction, increase negative)
    10 traces:      17.25 (      0.00)
   100 traces:      57.15 (      0.00)
  1000 traces:      93.89 (      0.00)
  2000 traces:      93.89 (      0.00)
```

While wall time has had a modest improvement:
```
Before(ThreadSafeBuffer) [operations/sec]
    10 spans:   74676.66
   100 spans:   11465.74
  1000 spans:    1208.54
  2000 spans:     486.27

After(CRubyTraceBuffer)  [operations/sec]
    10 spans:   82571.16
   100 spans:   13084.07
  1000 spans:    1397.91
  2000 spans:     512.44

Comparison (% faster; slower if negative)
    10 spans:      10.57
   100 spans:      14.11
  1000 spans:      15.67
  2000 spans:       5.38

Process finished with exit code 0
```

### CODE REVIEW

File: lib/ddtrace/buffer.rb [hunk 1]
```diff
@@ -72,48 +126,103 @@ def close
...
     end
+  end
+
+  # Trace buffer that stores application traces and
+  # can be safely used concurrently with CRuby.
+  #
+  # Under extreme concurrency scenarios, this class can exceed
+  # its +max_size+ by up to 4%.
+  #
+  # Because singular +Array+ operations are thread-safe in CRuby,
```
Comment (ericmustin) [inline lib/ddtrace/buffer.rb:137]: Is there a link to which Array operations are thread-safe in CRuby, or link to code? Might be something worth having in the comment to protect ourselves in the future

File: lib/ddtrace/buffer.rb [hunk 2]
```diff
@@ -2,23 +2,77 @@
...
-    def initialize(max_size)
-      @max_size = max_size
+  # Aggregate metrics:
+  # They reflect buffer activity since last #pop.
+  # These may not be as accurate or as granular, but they
+  # don't use as much network traffic as live stats.
+  class MeasuredBuffer
+    def initialize
+      @buffer_accepted = 0
+      @buffer_accepted_lengths = 0
```
Comment (ericmustin) [inline lib/ddtrace/buffer.rb:15]: I'm not sure I understand the difference between `@buffer_accepted_lengths` and `@buffer_spans` ? Are one of these redundant?

File: lib/ddtrace/buffer.rb [hunk 3]
```diff
@@ -72,48 +126,103 @@ def close
...
+  # we can implement the trace buffer without an explicit lock,
+  # while making the compromise of allowing the buffer to go
+  # over its maximum limit under extreme circumstances.
+  #
+  # On the following scenario:
+  # * 4.5 million spans/second.
+  # * Pushed into a single CRubyTraceBuffer from 1000 threads.
+  # The buffer can exceed its maximum size by no more than 4%.
+  #
+  # This implementation allocates 17-93% less memory and
```
Comment (ericmustin) [inline lib/ddtrace/buffer.rb:147]: Not a big deal but do we really want to add perf numbers in code comments? it's also a very broad range so i'm not sure how useful this comment is, maybe we can just say `'This implementation allocates significantly less memory and has modest speedup compared to Datadog::ThreadSafeBuffer'` or basically, something that won't be used against us in the feature 😅

File: lib/ddtrace/buffer.rb [hunk 4]
```diff
@@ -72,48 +126,103 @@ def close
...
+      else
+        # we should replace a random trace with the new one
+        replace_index = rand(len)
+        replaced_trace = @traces.delete_at(replace_index)
+        @traces << trace
+
+        # Check if we deleted the element right when the buffer
+        # was popped. In that case we didn't actually delete anything,
+        # we just inserted into a newly cleared buffer instead.
+        measure_drop(replaced_trace) if replaced_trace
```
Comment (ericmustin) [inline lib/ddtrace/buffer.rb:178]: is this a realistic scenario that we should be defensive about, ie, not insert or attempt to delete again? Or is it basically the thing that occurs 4% of the time in extreme conditions?

File: lib/ddtrace/ext/runtime.rb [hunk 5]
```diff
@@ -7,6 +7,7 @@ module Runtime
       LANG = 'ruby'.freeze
       LANG_INTERPRETER = (RUBY_ENGINE + '-' + RUBY_PLATFORM).freeze
       LANG_VERSION = RUBY_VERSION
+      RUBY_ENGINE =  ::RUBY_ENGINE # e.g. 'ruby', 'jruby', 'truffleruby'
```
Comment (ericmustin) [inline lib/ddtrace/ext/runtime.rb:10]: will this blowup anything in the frontend UI that relies on the `ruby` tag?

**Review by ericmustin — COMMENTED** [review]
I have a few small questions and nits, generally fine approving just want to be careful around the tag renaming stuff...this is great work, thanks @marcotc

File: lib/ddtrace/buffer.rb (hunk 1 above)
Comment (marcotc) [reply lib/ddtrace/buffer.rb:137]: Yes, it's in the `@see` section of this same documentation block, where it links to concurrent-ruby.

File: lib/ddtrace/buffer.rb (hunk 2 above)
Comment (marcotc) [reply lib/ddtrace/buffer.rb:15]: One is the number of traces pushed, the other is the total number of spans pushed. One tracer == one or more spans.

File: lib/ddtrace/buffer.rb (hunk 3 above)
Comment (marcotc) [reply lib/ddtrace/buffer.rb:147]: I was trying to ensure I had supportive arguments documented, but hard numbers are sure to be variable across any execution environment.

I want to make sure to capture information that allows for future decision making regarding trade-offs being taken here.

I'm ambivalent in keeping or removing the numbers. What do you think @brettlangdon?

File: lib/ddtrace/buffer.rb (hunk 4 above)
Comment (marcotc) [reply lib/ddtrace/buffer.rb:178]: This happens when the consumer has consumed all elements from the buffer right when we tried to access it.
We'll end up with a nil `replaced_trace`, which we can't measure.

This _can_ happen very frequently, as we consume from the buffer every few seconds, but only on a full buffer. This is not what causes the buffer to go over the limit.

The part the causes it to go over the limit is the time gap between us checking `@traces.length` and actually inserting the element.

File: lib/ddtrace/ext/runtime.rb (hunk 5 above)
Comment (marcotc) [reply lib/ddtrace/ext/runtime.rb:10]: Not sure I follow, `RUBY_ENGINE` is being used to decide which implementation to use by default here: https://github.com/DataDog/dd-trace-rb/pull/1172/files#diff-6650ecd3c000f2a31d4f5335841789a8R213-R217

That it's only use, and it's never exposed anywhere in the tracer.

File: lib/ddtrace/buffer.rb (hunk 2 above)
Comment (ericmustin) [reply lib/ddtrace/buffer.rb:15]: if `@buffer_accepted_lengths` is the number of traces pushed, then i believe we shouldn't be subtracting by the trace.length in `measure_drop` and `measure_accept`, right? If i'm off base here feel free to disregard

https://github.com/DataDog/dd-trace-rb/blob/7511b140aa973bc99572ebd080b13d0cc9bd1ba4/lib/ddtrace/buffer.rb#L30-L31

File: lib/ddtrace/ext/runtime.rb (hunk 5 above)
Comment (ericmustin) [reply lib/ddtrace/ext/runtime.rb:10]: derp, i got mixed up with `LANG`...disregard 🤦

File: lib/ddtrace/buffer.rb (hunk 4 above)
Comment (ericmustin) [reply lib/ddtrace/buffer.rb:178]: 👍  gotcha, seems fine then, appreciate the explanation

**Review by ericmustin — APPROVED** [review]
I'm still confused on the `@buffer_accepted_lengths` bit tbh but doesn't seem blocking to me, same for the code comment nit, so deferring to your best judgement on both. nice work!

File: lib/ddtrace/buffer.rb (hunk 2 above)
Comment (marcotc) [reply lib/ddtrace/buffer.rb:15]: 🐛 alert: ` @buffer_accepted_lengths -= trace.length `
Yeah, this is wrong, nice catch! It's an oldie too!

File: lib/ddtrace/buffer.rb (hunk 3 above)
Comment (brettlangdon) [reply lib/ddtrace/buffer.rb:147]: I love the transparency, but this code comment could become stale pretty quickly.

I agree with Eric, going with a simpler comment to ensure people are considerate of performance impact whenever they modify the code is 👍🏻 and then we can think about how to expose this metric in a different way.

Comment (marcotc) [reply lib/ddtrace/buffer.rb]: Thank you guys, I removed hard numbers from the comment.
There's a link to benchmarks than can be run at any time to validate if the stated performance gains still hold in this comment block already, so I'm thinking that covers the hard-numbers part.

File: lib/ddtrace/buffer.rb (hunk 2 above)
Comment (marcotc) [reply lib/ddtrace/buffer.rb:15]: I extracted this fix to a separate PR, https://github.com/DataDog/dd-trace-rb/pull/1182, as we'll likely want to report in separately in the release changelog.

**Review by ericmustin — APPROVED** [review]
🚀
